File: app/lib/features/contacts/contact_service.dart
import 'package:flutter_contacts/flutter_contacts.dart';

/// Result of writing one member to the phone book.
class ContactWriteResult {
  const ContactWriteResult({required this.success, this.deviceContactId, this.error});
  final bool success;
  final String? deviceContactId;
  final String? error;
}

/// Wraps the native contacts API (flutter_contacts 2.x). Contacts created by the
/// app are tagged with organisation "SBC" so they're identifiable for later
/// synchronisation and dedup (cahier §14/§15). No auto-messaging — ever.
class ContactService {
  Future<bool> requestPermission() async {
    final status = await FlutterContacts.permissions.request(PermissionType.readWrite);
    return _isUsable(status);
  }

  /// Current status without prompting — lets the UI explain itself before the
  /// system dialog appears, and lets it stay quiet once access is granted.
  Future<PermissionStatus> checkPermission() =>
      FlutterContacts.permissions.check(PermissionType.readWrite);

  Future<bool> hasPermission() async => _isUsable(await checkPermission());

  /// iOS 18 "limited" means the member picked specific contacts to share; that
  /// is still enough to create and dedup, so treat it as usable.
  static bool _isUsable(PermissionStatus s) =>
      s == PermissionStatus.granted || s == PermissionStatus.limited;

  /// True if a contact with this phone number already exists on the device
  /// (cahier §15 duplicate detection — the authoritative check is native).
  Future<bool> existsByPhone(String phone) async {
    final normalized = _digits(phone);
    if (normalized.isEmpty) return false;
    final contacts = await FlutterContacts.getAll(properties: {ContactProperty.phone});
    for (final c in contacts) {
      for (final p in c.phones) {
        final d = _digits(p.number);
        if (d.isNotEmpty && (d.endsWith(normalized) || normalized.endsWith(d))) {
          return true;
        }
      }
    }
    return false;
  }

  Future<ContactWriteResult> addSbcContact({
    required String firstName,
    String? lastName,
    String? phone,
    String? profession,
    String? city,
    String? country,
  }) async {
    try {
      // Append an "SBC" suffix to the display name so users can recognise which
      // contacts came from SBC. The suffix rides on the last-name field so the
      // phone shows "Prénom Nom SBC" (or just "Membre SBC" when nameless), with
      // exactly one suffix and no double spaces.
      final suffixedLast = _withSbcSuffix(firstName: firstName, lastName: lastName);
      final contact = Contact(
        name: Name(first: firstName, last: suffixedLast),
        phones: [if (phone != null && phone.isNotEmpty) Phone(number: phone)],
        organizations: [Organization(name: 'SBC', jobTitle: profession ?? '')],
      );
      final id = await FlutterContacts.create(contact);
      return ContactWriteResult(success: true, deviceContactId: id);
    } catch (e) {
      return ContactWriteResult(success: false, error: e.toString());
    }
  }

  String _digits(String s) => s.replaceAll(RegExp('[^0-9]'), '');

  /// Builds the last-name field so the full display name ends with exactly one
  /// " SBC". If both names are empty the whole contact reads "Membre SBC"
  /// (first name empty, last name "Membre SBC").
  static String _withSbcSuffix({required String firstName, String? lastName}) {
    final first = firstName.trim();
    final last = (lastName ?? '').trim();
    if (first.isEmpty && last.isEmpty) return 'Membre SBC';
    return '$last SBC'.trim();
  }
}

File: ios/SBCContacts/Features/Contacts/ContactService.swift
import Contacts
import Foundation

/// Result of writing one member to the phone book.
struct ContactWriteResult: Sendable {
    let success: Bool
    var deviceContactId: String?
    var error: String?
}

/// Wraps the native Contacts framework. Contacts created by the app are tagged
/// with organisation "SBC" so they're identifiable for later synchronisation
/// and dedup (cahier §14/§15). No auto-messaging — ever.
struct ContactService: Sendable {
    enum Access: Sendable {
        case notDetermined, granted, denied
    }

    /// Current status without prompting — lets the UI explain itself before the
    /// system dialog appears, and stay quiet once access is granted.
    func checkPermission() -> Access {
        Self.map(CNContactStore.authorizationStatus(for: .contacts))
    }

    /// iOS 18 "limited" means the member picked specific contacts to share;
    /// that is still enough to create and dedup, so it counts as granted.
    private static func map(_ status: CNAuthorizationStatus) -> Access {
        switch status {
        case .authorized: return .granted
        case .notDetermined: return .notDetermined
        case .denied, .restricted: return .denied
        default:
            if #available(iOS 18.0, *), status == .limited { return .granted }
            return .denied
        }
    }

    func requestPermission() async -> Bool {
        switch checkPermission() {
        case .granted: return true
        case .denied: return false
        case .notDetermined:
            _ = try? await CNContactStore().requestAccess(for: .contacts)
            return checkPermission() == .granted
        }
    }

    /// True if a contact with this phone number already exists on the device
    /// (cahier §15 duplicate detection — the authoritative check is native).
    func existsByPhone(_ phone: String) async -> Bool {
        let normalized = Self.digits(phone)
        guard !normalized.isEmpty else { return false }
        let request = CNContactFetchRequest(keysToFetch: [CNContactPhoneNumbersKey as CNKeyDescriptor])
        var found = false
        try? CNContactStore().enumerateContacts(with: request) { contact, stop in
            for number in contact.phoneNumbers {
                let d = Self.digits(number.value.stringValue)
                if !d.isEmpty, d.hasSuffix(normalized) || normalized.hasSuffix(d) {
                    found = true
                    stop.pointee = true
                    return
                }
            }
        }
        return found
    }

    /// Writes one member to the phone book.
    ///
    /// `displayName` is the member's whole name and the only name input: SBC
    /// puts the full name in `name` and often leaves `firstName` empty, so
    /// passing both wrote "Claude Durel Claude Durel". Splitting one string
    /// here is the only way the two halves cannot disagree.
    func addSBCContact(
        displayName: String,
        phone: String?,
        profession: String?
    ) async -> ContactWriteResult {
        let contact = CNMutableContact()
        let name = Self.nameParts(displayName)
        contact.givenName = name.given
        contact.familyName = name.family
        if let phone, !phone.isEmpty {
            contact.phoneNumbers = [CNLabeledValue(label: CNLabelPhoneNumberMobile, value: CNPhoneNumber(stringValue: phone))]
        }
        contact.organizationName = "SBC"
        contact.jobTitle = profession ?? ""

        let request = CNSaveRequest()
        request.add(contact, toContainerWithIdentifier: nil)
        do {
            try CNContactStore().execute(request)
            return ContactWriteResult(success: true, deviceContactId: contact.identifier)
        } catch {
            return ContactWriteResult(success: false, error: error.localizedDescription)
        }
    }

    /// First word → given name, the rest → family name with "SBC" appended, so
    /// the phone's own list shows "Claude Durel SBC" and every contact the app
    /// wrote is recognisable without opening it. Idempotent on "… SBC".
    static func nameParts(_ displayName: String) -> (given: String, family: String) {
        let parts = displayName.split(whereSeparator: \.isWhitespace).map(String.init)
        guard let first = parts.first else { return ("Membre", "SBC") }
        let rest = parts.dropFirst().joined(separator: " ")
        let family: String
        if rest.isEmpty {
            family = "SBC"
        } else if rest.uppercased().hasSuffix("SBC") {
            family = rest
        } else {
            family = "\(rest) SBC"
        }
        return (first, family)
    }

    private static func digits(_ s: String) -> String {
        s.filter(\.isNumber)
    }
}
